### PR TITLE
Reorganization of AST module into declare and define steps.

### PR DESCRIPTION
**Pull request by adjivas**
Reorganization of AST module into _declare_ and _define_ steps.

### CODE REVIEW

File: telamon-gen/src/ast/typing.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
+use super::*;
+use std::ops::Deref;
```
Comment (adjivas) [inline telamon-gen/src/ast/typing.rs]: rename module into typing_context

File: telamon-gen/src/ast/mod.rs [hunk 2]
```diff
@@ -524,6 +430,24 @@ pub enum EnumStatement {
     AntiSymmetric(Spanned<Vec<(String, String)>>),
 }
 
+impl EnumStatement {
```
Comment (ulysseB) [inline telamon-gen/src/ast/mod.rs]: the should be in the enum typing file

**Comment by adjivas** [comment]
@ulysseB Ready for review again.

**Comment by adjivas** [comment]
bors try

File: telamon-gen/src/ast/choice/enumeration.rs [hunk 3]
```diff
@@ -148,6 +105,49 @@ impl EnumDef {
         Ok(())
     }
 
+    /// This checks the undefined of value or alias from alias or antisymmetric.
```
Comment (ulysseB) [inline telamon-gen/src/ast/choice/enumeration.rs]: Replace with "Checks if the values referenced in `EnumStatement`s are defined"

File: telamon-gen/src/ast/typing_context.rs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                object_kind: pre,
+                object_name,
+            })
+        } else {
+            Ok(())
+        }
+    }
+
+    /// This checks the undefined of SetDef superset and arg.
+    pub fn check_set_define(
```
[the remark this answers is not in the document]
Comment (ulysseB) [reply telamon-gen/src/ast/typing_context.rs]: Ping me when you think you know what to do, just to make sure we are one the same line.

**Comment by adjivas** [comment]
@ulysseB Ready for review again.

File: telamon-gen/src/ast/typing_context.rs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            Err(TypeError::Redefinition {
+                object_kind: pre,
+                object_name,
+            })
+        } else {
+            Ok(())
+        }
+    }
+
+    /// This checks the undefined of SetDef superset and arg.
```
Comment (ulysseB) [inline telamon-gen/src/ast/typing_context.rs]: rephrase: Check if the referenced set is defined

File: telamon-gen/src/ast/typing_context.rs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+/// CheckContext is a type system.
+#[derive(Debug, Default)]
+pub struct CheckerContext {
+    /// Map Name of unique identifiant.
+    hash_set: HashMap<String, Spanned<Hint>>,
+    hash_choice: HashMap<String, Spanned<Hint>>,
+}
+
+impl CheckerContext {
+    /// This checks the redefinition of SetDef.
```
Comment (ulysseB) [inline telamon-gen/src/ast/typing_context.rs]: rephrase: Declares a set and ensures it is not defined twice

File: telamon-gen/src/ast/typing_context.rs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            Err(TypeError::Redefinition {
+                object_kind: pre,
+                object_name,
+            })
+        } else {
+            Ok(())
+        }
+    }
+
+    /// This checks the redefinition of ChoiceDef (EnumDef and IntegerDef).
```
Comment (ulysseB) [inline telamon-gen/src/ast/typing_context.rs]: rephrase: Declares a choice and ensures it is not defined twice

**Pull request merged**